pilot/pkg/proxy/envoy/v2/datastructures.go

```diff
@@ -0,0 +1,139 @@
+//    catch all TCPRoute
+// 6. Every HTTP port of a service will share the same route configuration (list of virtual hosts)
+//    until further delineation is created by route rules.
+// 7. Similarly, every TCP port of a service will share the same TCPRoute.
+// 7. Routes change (add/delete/update) typically when route rules change
+// 9. Routes can also change when ports are added/removed from the service
+
+// Listeners (react to updates in services, gateways, ext services)
+// 1. Every unique port in the system will have an associated listener.
+// 2. A service can generate one or more listeners based on the number of ports it listens on
+// 3. New listeners are added when a service (incl gateway/external service) with unique
+//    (previously unused) set of ports are added.
+// 4. Listeners are updated when one or more services/gateways/extsvs are added/removed to/from
+//    the same listener port.
+// 5. Listeners are removed when there are no services/gateways/extsvcs listening on the listener
+//    port.
+
+// The indices here are doubly linked. A pair of objects are dependent on each other if they have
+// to be updated when one changes.
+
+// BoxedVirtualHost holds the virtual host proto representation and pointers to
+// owning route configurations.
+type BoxedVirtualHost struct {
+	// The proto representation of this virtual host
+	protoVirtualHost *xdsapi.VirtualHost
+	// list of route configurations (RDS responses) that hold this virtual host
+	routeConfigurations []*BoxedRouteConfiguration
+}
+
+// BoxedRouteConfiguration holds the route configuration proto representation and pointers to
+// owning listeners
+type BoxedRouteConfiguration struct {
+	// The proto representation of this route configuration
+	protoRouteConfiguration *xdsapi.RouteConfiguration
+	// list of virtual hosts attached to this route configuration
+	virtualHosts []*BoxedVirtualHost
+	// list of listeners that use this route configuration
+	listeners []*xdsapi.Listener
+}
+
+// BoxedEndpoint holds the endpoint proto representation and pointers to
+// owning clusters
+type BoxedEndpoint struct {
+	// The proto representation of this endpoint
+	protoEndpoint *xdsapi.LbEndpoint
+	// list of clusters that contain this endpoint
+	clusters []*BoxedCluster
+	// list of labels associated with this endpoint
+	labels []*model.Labels
+}
+
+// BoxedCluster holds the cluster proto representation, pointers to endpoints
+type BoxedCluster struct {
+	// The proto representation of this cluster
+	protoCluster *xdsapi.Cluster
+	// list of endpoints associated with this cluster
+	endpoints []*BoxedEndpoint
+	// List of labels associated with this cluster
+	labels []*model.Labels
+}
+
+// Indices is a set of reverse indices to quickly locate endpoints, clusters, routes
+// and listeners
+type Indices struct {
+	// Mutex guards guarantees consistency of updates to members shared across threads.
+	mu sync.RWMutex
```

Comment:
global mutex per all-nodes seems to be a bottleneck.
How do we plan to stage config rollout here? Shouldn't we keep both old and new configs for some time?

Reply:
It's not global if per node. 

Not sure what is best - and don't remember memory cost per mutex, but usually easier to move to global lock than reverse.

Reply:
Ignore the mutex please. It was a copy paste thing from an old struct definition. I wouldn’t have such a coarse grained mutex for this entire struct.